### DISCUSSION TITLE
Wagmi v2

### DISCUSSION THREAD

**Discussion by tmm** [wevm/wagmi#3068] [Announcements]
Very excited to announce the next major version of Wagmi!

Wagmi v2 redesigns the core APIs to mesh better with [Viem](https://viem.sh) and [TanStack Query](https://tanstack.com/query/v5/docs/react). This major version transforms Wagmi into a light wrapper around these libraries, sprinkling in multichain support and account management. As such, there are some breaking changes and deprecations. Check out the [migration guide](https://rc.wagmi.sh/react/guides/migrate-from-v1-to-v2) for more info.

There are also some incredible features included in this release, like:

- Full TanStack Query support + queryKey exports
- Connect multiple connectors at a time
- Switch chains while disconnected
- EIP-6963 enabled
- Strongly typed chainId and chain properties
- Smaller bundle size and improved tree-shaking
- Updated docs at https://rc.wagmi.sh and NPM provenance

We recommend trying it out in your projects, but there may be breaking changes before the final release. If you find bugs or have feedback, please [open an issue](https://github.com/wagmi-dev/wagmi/issues/new/choose) or reply to this discussion. All development is happening on the [`rc`](https://github.com/wagmi-dev/wagmi/tree/rc) branch.

In general, since major versions don't happen that often, feel free to give feedback on anything. We appreciate the support!

**Comment by DannyDelott** [comment]
Excellent work on the release. Migration has been pretty smooth.

One issue: We get our `connectors` from RainbowKit. Is there a workaround to convert a v1 wagmi connector to a v2 one? I suspect Rainbowkit isn't far behind, but perhaps there's a simple solution in the meantime.

Example snippet:

```ts
import { createConfig, http } from "wagmi";
import { mainnet } from "wagmi/chains";
import { getDefaultWallets } from "@rainbow-me/rainbowkit";

const { connectors } = getDefaultWallets({
  appName: "my app",
  projectId: 'project-id'
  chains: [mainnet],
});

export const wagmiConfig = createConfig({
  chains: [mainnet]
  connectors, // <-- incompatable type
  transports: {
    [mainnet.id]: http('alchemy-url'),
  },
});
```

https://www.rainbowkit.com/docs/installation#configure

**Reply by jasperverbeet** [reply]
Looks like RainbowKit is working on it: https://github.com/rainbow-me/rainbowkit/pull/1601.

We ended up building their PR and patching the dist folder locally (using pnpm patch). Works like a charm.

**Reply by mhpaler** [reply]
We're interested in doing something similar. Can you expand a bit more on how you performed this?

**Comment by oxSaturn** [comment]
Is `react` plugin from `@wagmi/cli/plugins` removed? Couldn't find it in source code, and it's not mentioned in the migration guide:

![CleanShot-2NuOmk0D@2x](https://github.com/wagmi-dev/wagmi/assets/126733611/348f0855-6571-46bd-b107-7c36ce274224)

**Reply by tmm** [reply]
It's still a work-in-progress!

**Reply by tmm** [reply]
It's live. 

https://beta.wagmi.sh/cli/api/plugins/react

**Comment by art-eth** [comment]
What is the estimated release date?

**Reply by tmm** [reply]
Depends on how much feedback folks give, etc. Alpha will probably be for next month, then beta for a month, then major version release. This could change, but it's how the core team is thinking about it right now.

**Reply by MartinPucan** [reply]
@tmm please, is there more accurate estimate of when the release could be? Thank you!

**Reply by tmm** [reply]
Release candidate this week. Stable release first week of January.

**Comment by oxSaturn** [comment]
Wondering if anyone run into this issue with `pnpm wagmi init`:

```
/node_modules/@wagmi/cli/dist/esm/commands/generate.js:5
import { ensureDir, writeFile } from 'fs-extra';
                    ^^^^^^^^^
SyntaxError: Named export 'writeFile' not found. The requested module 'fs-extra' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'fs-extra';
const { ensureDir, writeFile } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)

Node.js v18.17.1
```

**Comment by script-money** [comment]
> Wondering if anyone run into this issue with `pnpm wagmi init`:
> 
> ```
> /node_modules/@wagmi/cli/dist/esm/commands/generate.js:5
> import { ensureDir, writeFile } from 'fs-extra';
>                     ^^^^^^^^^
> SyntaxError: Named export 'writeFile' not found. The requested module 'fs-extra' is a CommonJS module, which may not support all module.exports as named exports.
> CommonJS modules can always be imported via the default export, for example using:
> 
> import pkg from 'fs-extra';
> const { ensureDir, writeFile } = pkg;
> 
>     at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
>     at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)
> 
> Node.js v18.17.1
> ```

I has some issue, have to change code in node_modules

**Reply by iceanddust** [reply]
Same here. Applying the suggestion in the error message in the node_modules code fixes the bug. Although a better, tree-shakable approach I think would be to upgrade `fs-extra` to v11 and change the import source to `fs-extra/esm`

**Comment by generatorpoint** [comment]
Any updates on the wagmi-cli react plugin? Will we have a chance to test it out soon?

**Reply by tmm** [reply]
Working on it shortly. Will reply to this thread (and [this one](https://github.com/wagmi-dev/wagmi/discussions/3068#discussioncomment-7140097)) once it's complete. Once it is out, there will be plenty of time for feedback in beta before heading to stable release.

**Reply by tmm** [reply]
It's released.

https://beta.wagmi.sh/cli/api/plugins/react

**Comment by unicape** [comment]
Are the `config` parameters in useConfig really useful?

Will anyone initialize the parameters externally and then pass them to useConfig to obtain them?
I don't think it makes any sense

**Reply by tmm** [reply]
It's probably not useful for most users, but for library authors (e.g. us building Wagmi) or folks that don't want to wrap their app in the Wagmi React Context Provider, it's useful. It streamlines the internal hook typing and [bypasses the missing React Context Provider error](https://github.com/wagmi-dev/wagmi/blob/beta/packages/react/src/hooks/useConfig.ts#L20).

**Comment by samwhosung** [comment]
Will v2 support async loading of connectors? https://github.com/wevm/wagmi/discussions/2402

**Reply by tmm** [reply]
`createConfig`'s [`connectors`](https://beta.wagmi.sh/react/api/createConfig#connectors) option is synchronous. If you don't want to configure connectors globally via `createConfig`, most hooks allow you to pass in a `connector` property.

**Reply by samwhosung** [reply]
@tmm does this also work with v1?

**Comment by unicape** [comment]
Can the v2 version add more examples of hooks? I think there are only a small part now.

**Reply by tmm** [reply]
Any specific examples you think would be helpful? Planning to port over [all of these](https://wagmi.sh/examples/connect-wallet) (except the SIWE example).

**Reply by unicape** [reply]
Thank you, that's enough

**Comment by firnprotocol** [comment]
great work.

i am completely unable to `switchChains` or `switchChainsAsync` in wagmi2. on all connectors (injected MetaMask, WalletConnect, etc.), i get the error:
```javascript
SwitchChainError: An error occurred when attempting to switch chain.

Details: Chain not configured.

Version: @wagmi/core@2.0.0-rc.0
Version: viem@2.0.0-rc.0
    at Object.switchChain (injected.js:274:1)
    at async switchChain (switchChain.js:14:1)
```
with `error.code === 4902`. thanks in advance

**Reply by tmm** [reply]
Can you create a bug with a minimal reproduction? Can't look into it without more info.

**Reply by firnprotocol** [reply]
what'd be very helpful is to have a "working minimal wagmi 2 repo" that i can fork in order to replicate. otherwise i need to set up a micro-react app, etc. would that be doable?

**Reply by tmm** [reply]
Can use [this CodeSandbox](https://stackblitz.com/edit/vitejs-vite-5dnjcb) for now. There are also a [few playgrounds](https://github.com/wevm/wagmi/tree/rc/playgrounds) in the repo you can use.

**Reply by firnprotocol** [reply]
1. i'm an idiot, i solved it on metamask by replacing ~`switchChain(chainId)`~ --> `switchChain({ chainId })`. this looks like a semantic change from wagmi 1.
2. on WalletConnect, i am actually now getting a new error:
```javascript
SwitchChainError: An error occurred when attempting to switch chain.

Details: Cannot set properties of undefined (setting 'defaultChain')
Version: viem@2.0.0-rc.0
    at Object.switchChain (walletConnect.js:184:1)
    at async switchChain (switchChain.js:14:1)
```
let me try to see if i can extend that sandbox to reproduce.

**Reply by firnprotocol** [reply]
k, was able to trivially reproduce just by adding the walletconnect connector to that sandbox: https://stackblitz.com/edit/vitejs-vite-ty17d7

steps to reproduce: connect using walletconnect (any wallet), then try switching chains

**Reply by tmm** [reply]
Can you create an issue so we can track it?

**Reply by firnprotocol** [reply]
k, see https://github.com/wevm/wagmi/issues/3347

**Reply by kristiegiles** [reply]
I ran into this, too. The `chainId` --> `{ chainId }` change isn't mentioned in the migration guide.

**Reply by tmm** [reply]
PR welcome!

**Comment by jasperverbeet** [comment]
It looks like V2 removed the `suspense: true` option, is this feature coming back in the future?

In v4 we could pass down the `suspense: true` property to `useQuery`, but now it looks like React Query has added a `useSuspenseQuery` hook in v5 (https://tanstack.com/query/latest/docs/react/reference/useSuspenseQuery).

**Reply by jasperverbeet** [reply]
One potential solution would be to update this function:

https://github.com/wevm/wagmi/blob/11936387bdd7db2a672ddcf9dee332828352f213/packages/react/src/utils/query.ts#L73C1-L85C1

To do something like:

```ts
// This line is new
const tanstack_useQueryFn = parameters.suspense
    ? tanstack_useSuspenseQuery
    : tanstack_useQuery;

  const result = tanstack_useQueryFn({
    ...(parameters as any),
    queryKeyHashFn: hashFn // for bigint support
  }) as UseQueryReturnType<data, error>;
  result.queryKey = parameters.queryKey;
  return result;
```

The following params are not available for `useSuspenseQuery` (see docs above), so these would have to be removed from the types:

- suspense
- throwOnError
- enabled
- placeholderData

**Reply by tmm** [reply]
TanStack Query v4 suspense support was [experimental](https://tanstack.com/query/v5/docs/react/guides/migrating-to-v5#new-hooks-for-suspense). Since the option was dropped from `useQuery`, it was also [removed from Wagmi](https://rc.wagmi.sh/react/guides/migrate-from-v1-to-v2#removed-suspense-property). Instead, you can use `useSuspenseQuery` along with TanStack Query-related exports from the `'wagmi/query'` entrypoint.

```ts
import { useSuspenseQuery } from '@tanstack/react-query'
import { useConfig } from 'wagmi'
import { getBalanceQueryOptions } from 'wagmi/query'

const config = useConfig()
const result = useSuspenseQuery(
  getBalanceQueryOptions(config, { address: '0x…' })
)
```

In the future, we might add separate suspense hooks (e.g. `useSuspenseBalance`, `useSuspenseReadContract`. `useSuspenseEnsName`), but no immediate plans at this time as it's more to manage and we haven't heard from folks that it is an important feature.

**Reply by jasperverbeet** [reply]
Thank you for your reply :).

Initially, we ended up reimplementing those hooks as you described, but having that overhead in our app became a big point of frustration. 

We later decided to patch it out since we use the new Suspense API everywhere.

```sh
pnpm patch wagmi@2.0.0-rc.0 
```

```diff
diff --git a/dist/esm/utils/query.js b/dist/esm/utils/query.js
index 4a08d4dc6796060752c48199a53f9798077e3079..74b5d19dfe73dc7c8980912e8c29840f8c8dddfb 100644
--- a/dist/esm/utils/query.js
+++ b/dist/esm/utils/query.js
@@ -1,4 +1,4 @@
-import { replaceEqualDeep, useInfiniteQuery as tanstack_useInfiniteQuery, useQuery as tanstack_useQuery, } from '@tanstack/react-query';
+import { replaceEqualDeep, useInfiniteQuery as tanstack_useInfiniteQuery, useSuspenseQuery as tanstack_useQuery, } from '@tanstack/react-query';
 import { deepEqual, } from '@wagmi/core/internal';
 import { hashFn } from '@wagmi/core/query';
 // Adding some basic customization.
diff --git a/src/utils/query.ts b/src/utils/query.ts
index 79dcb580a846610c8e059ce6adec2370396df996..90662e74a2284c0ec5be5f4866f5a40f03d62fce 100644
--- a/src/utils/query.ts
+++ b/src/utils/query.ts
@@ -9,7 +9,7 @@ import {
   type UseQueryResult,
   replaceEqualDeep,
   useInfiniteQuery as tanstack_useInfiniteQuery,
-  useQuery as tanstack_useQuery,
+  useSuspenseQuery as tanstack_useQuery,
 } from '@tanstack/react-query'
 import {
   type Evaluate,
```

I hope we get this feature natively in the future, but other than that, thank you for the hard work on this project. It's helped us massively over the last 1.5y or so.

**Reply by nicosampler** [reply]
@tmm could it be possible to add a `wagmi config` to indicate using `useSuspenseQuery` instead of `useQuery` as @jasperverbeet is suggesting?

**Comment by priyeshkadbe** [comment]
hey devs...
can anyone help me out... I am fetching gasPrice using gasFeeData hook ... it gives correct gasPrice for polygon and arbitrum chain.. but for BNB chain it is returning incrorrect e.g on BSC Scan the gasPrice is shouwing 3Gwei .. but the gasFeeData returns 9Gwei

**Comment by kristiegiles** [comment]
<img width="548" alt="Screenshot 2024-01-03 at 9 39 35 AM" src="https://github.com/wevm/wagmi/assets/25679998/4689ee0c-bf2f-4d6c-b944-0e9fe7b4aaff">

[https://stackblitz.com/edit/vitejs-vite-9e1qqk?file=src%2Fcomponents%2FConnect.tsx](https://stackblitz.com/edit/vitejs-vite-9e1qqk?file=src%2Fcomponents%2FConnect.tsx
)
Why are there two separate MetaMask connectors now? It seems one has an id of "metaMaskSDK" and one is "io.metamask."

**Reply by firnprotocol** [reply]
i encountered this too. one is being injected by MetaMask and is being automatically detected; the other corresponds to an explicit `metaMask()` connector.

the solution I found was to remove the explicit connector and rely only on the injected one.

**Reply by tmm** [reply]
`metaMask` connector supports mobile MetaMask. `injected` supports injected wallets, which MetaMask so happens to fall under (similar to how the `coinbaseWallet` connector is cross platform and the Coinbase Wallet extension also works with `injected`). For mobile support without using the `metaMask` connector, you can rely on the `walletConnect` connector.

**Reply by kristiegiles** [reply]
I like the idea of supporting mobile MetaMask via its own button, so I tried filtering out the "io.metamask" provider before displaying the buttons and leaving the SDK one.

It seems that if I log in with the SDK one, after a period of time on a stale tab, it switches itself to using the "io.metamask" one. 

Then if I try to make a tx with that connector (or call methods on the walletClient), I get this undefined mobile provider error: 
<img width="623" alt="Screenshot 2024-01-05 at 3 19 20 PM" src="https://github.com/wevm/wagmi/assets/25679998/0b705052-0950-484c-9e48-6b174fef6198">

It works fine with the SDK connector sometimes, but I am sometimes getting the above error.

**Reply by glitch-txs** [reply]
The last button is actually from [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963), maybe it could be filtered when MetaMask connector is present? AFAIK I heard MM would also handle EIP-6963 in their SDK, but I'm not sure if they do.
You can verify this by setting [multiInjectedProviderDiscovery](https://wagmi.sh/react/api/createConfig#multiinjectedproviderdiscovery) to false, it detects your installed extensions. `io.metamask` is the reverse DNS from the provider.

**Reply by glitch-txs** [reply]
For the SDK errors I'd try testing it in standalone to check if it's an issue on Wagmi or the SDK side

**Reply by glitch-txs** [reply]
Seems like MM SDK is already [using EIP-6963](https://github.com/MetaMask/metamask-sdk/blob/14dffccfa61ecb9bf89a687d5acca12f74b5a996/packages/sdk/src/utils/get-browser-extension.ts#L19), should we change its ID to `io.metamask`? would this filter the provider to avoid duplicates?

**Reply by tmm** [reply]
Could be up for this depending out the behavior. If someone wants to sketch out a PR, can take a look.

**Reply by glitch-txs** [reply]
my bad, the filter lives inside the mipd but not globaly for all connectors, so changing the ID doesn't work, I guess the best approach is to filter them manually outside Wagmi?

**Reply by androidsixteen** [reply]
I'm running into the same issue that @kristiegiles screenshotted above:
`SDK state invalid -- undefined mobile provider`

I've set`multiInjectedProviderDiscovery` to false and have explicitly added MetaMask as a connector, so would prefer to get rid of this issue using those settings. The downside of relying on the injected connector is that you can't always assume that the user will have MetaMask installed

Was there any resolution for this, or any workaround we can use to suppress the error?

**Reply by jxom** [reply]
This is a MetaMask SDK issue unfortunately. Will ping them.

**Comment by kristiegiles** [comment]
I have a Next.js app that uses the pages directory. I had previously disabled SSR with dynamic imports on the wagmi config and any components using wagmi.  For example: 

```
const WagmiConfig = dynamic(
  import('wagmi').then((mod) => mod.WagmiConfig),
  { ssr: false }
);
```

I tried changing those to normal imports and adding `ssr: true` to the config. For example:
```
import { WagmiProvider } from 'wagmi';

const config = createConfig({
  ssr: true,
}
``` 

But the hydration errors came back...is this expected?

<img width="649" alt="Screenshot 2024-01-03 at 10 50 21 AM" src="https://github.com/wevm/wagmi/assets/25679998/a574ee8d-5cbd-4e12-9e37-f27c16537b15">

**Reply by tmm** [reply]
Can you provide a minimal reproduction? Hard to tell if this is related to Wagmi or something else without it.

**Reply by kristiegiles** [reply]
Here is a minimal reproduction: https://github.com/kristiegiles/nextjs-wagmi-qur2tj

The hydration mismatch error does not show in StackBlitz but should show if you run it locally. 

Changing the import to 
```
import dynamic from 'next/dynamic'
const WagmiProvider = dynamic(
  import("wagmi").then((mod) => mod.WagmiProvider),
  { ssr: false }
);
```
makes the hydration error go away.

**Reply by kristiegiles** [reply]
After upgrading to the latest version, I no longer get this error in my minimal reproduction.

**Comment by iceanddust** [comment]
Are you planning to add back to the `actions` and `react` plugins of the CLI package the ability to select which hooks and functions we want to be generated? I have a project where we make use of this functionality in v1 like so:
```ts
...
plugins: [
    react({
        useContractEvent: false,
        useContractRead: true,
        useContractFunctionRead: false,
        ...
    }),
    ...
],
... 
```

**Reply by tmm** [reply]
Not at the moment as any unused code is tree-shakable away and it's not work the extra flags IMO.

**Reply by iceanddust** [reply]
Fair enough.
I'm trying to generate the new file, but I'm running into some errors with the watch contract event generated hooks. It's claiming that `functionName` is not a property of `createUseWatchContractEvent`. I think it's because [this line](https://github.com/wevm/wagmi/blob/main/packages/cli/src/plugins/react.ts#L229C71-L229C99) in the react CLI plugin should use `eventName` instead of `functionName`

**Reply by tmm** [reply]
Nice spot! PR welcome or create an issue and I can fix shortly.

**Reply by iceanddust** [reply]
Cool, I just opened a PR for that :)

**Reply by tmm** [reply]
Thank you! Will release soon.

**Comment by kristiegiles** [comment]
<img width="519" alt="Screenshot 2024-01-05 at 3 41 15 PM" src="https://github.com/wevm/wagmi/assets/25679998/c38dac44-b276-48e4-a32d-b95e9c44d479">

The migration guide shows data?.gas but in my version (2.0.0-rc.2), data from useEstimateGas is a bigint and data?.gas is undefined.

**Reply by kristiegiles** [reply]
Updated migration guide

**Comment by hiromaily** [comment]
Hi Wagmi team. This v2 is awesome!
I'm still using v1 and am curious about when wagmi v1 maintenance is end.

**Reply by tmm** [reply]
V1 will receive critical security fixes until the majority of users move over to v2.

**Comment by bjfresh** [comment]
Curious as to whether useReadContract/useWriteContract/useWaitForTransactionReceipt might eventually (re)gain support for onError/onSuccess callbacks from v1? These were especially useful eg. for Sentry error logging on failures. There are alternate ways of accomplishing this using isError/error, of course, but the previous interface was really nicely encapsulated.

**Reply by tmm** [reply]
`useWriteContract` [still has these callbacks](https://wagmi.sh/react/api/hooks/useWriteContract#onsuccess). They were removed from [query hooks](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#callbacks-on-usequery-and-queryobserver-have-been-removed), not mutation hooks. Unlikely they come back for query hooks unless TanStack Query adds them back (which seems extremely unlikely).

**Comment by 0xMySt1c** [comment]
is there a reason you cannot pass the data.request directly to the useWriteContract hook.

I wanted to bundle the entire useSimulateContract -> useWriteContract lifecycle into a custom hook (made the error handling much cleaner on the component side), but because useWriteContract no longer returns the write functions not sure it is possible?

**Reply by tmm** [reply]
It's not how [TanStack Query](https://github.com/TanStack/query) does mutations. Wagmi has used TanStack Query internally since v0.2. Wagmi v2 aligns closer to how TanStack Query works so runtime and type-level behavior is more robust and stable.
